compiler/rustc_middle/src/values.rs

```diff
@@ -131,6 +131,27 @@ impl<'tcx> Value<TyCtxt<'tcx>> for ty::EarlyBinder<ty::Binder<'_, ty::FnSig<'_>>
     }
 }
 
+impl<'tcx> Value<TyCtxt<'tcx>> for &[ty::Variance] {
+    fn from_cycle_error(
+        tcx: TyCtxt<'tcx>,
+        cycle_error: &CycleError,
+        _guar: ErrorGuaranteed,
+    ) -> Self {
+        if let Some(frame) = cycle_error.cycle.get(0)
+            && frame.query.dep_kind == dep_kinds::variances_of
+            && let Some(def_id) = frame.query.def_id
+        {
+            let n = tcx.generics_of(def_id).params.len();
+            vec![ty::Variance::Bivariant; n].leak()
```

Comment:
Could you explain what happens here?
When will this impl be called? Is this logic for error recovery?

(Also `tcx.arena.allocate` could probably be used instead of `leak`.)

Reply:
> When will this impl be called? Is this logic for error recovery?

This is not part of recovery, but of `cycle_delay_bug`. In order to hide cycle errors in favor of more useful errors, we can make the query cycle turn into a `span_delayed_bug`. But then we need a value to return from the query, which is where all the impls in this module are invoked. They produce a dummy value for the query.

> (Also `tcx.arena.allocate` could probably be used instead of `leak`.)

Then I would have to transmute the result 😆 

I guess if someone is running rustc in a loop from within the same process, and keep hitting this cycle error, they would slowly lose memory

Reply:
edit: leak is already being used elsewhere in this file, so this is more of a general point that we could figure out, but it seems low priority, so I think just avoiding the unsafe code is best

Reply:
>This is not part of recovery

I mean `ErrorGuaranteed` is passed to this function, so some error already happened (or at least was stashed)?
So this dummy value is never created on a good path.

Reply:
`span_delayed_bug` creates an `ErrorGuaranteed` without emitting an error. It will ICE the compiler if you forgot to emit an error later